tests/sentry/api/endpoints/test_project_index.py

```diff
@@ -158,3 +160,51 @@ def test_deleted_token_with_internal_integration(self):
         path = reverse(self.endpoint)
         response = self.client.get(path, HTTP_AUTHORIZATION=f"Bearer {token}")
         assert response.status_code == 401
+
+    def test_valid_with_public_integration(self):
+        project = self.create_project(organization=self.organization, teams=[self.team])
```

Comment:
This repeated code should be pulled into a method to make the tests easier to read:
```py
def create_sentry_app_token(self) -> ApiToken:
    sentry_app = self.create_sentry_app(
        scopes=("project:read",),
        published=False,
        verify_install=False,
        name="Super Awesome App",
    )
    installation = self.create_sentry_app_installation(
        slug=sentry_app.slug, organization=self.organization, user=self.user
    )

    return GrantExchanger.run(
        install=installation,
        code=installation.api_grant.code,
        client_id=sentry_app.application.client_id,
        user=sentry_app.proxy_user,
    )

def test_valid_with_public_integration(self):
    token = self.create_sentry_app_token()
    ...

def test_deleted_token_with_public_integration(self):
    token = self.create_sentry_app_token()
    token.delete()
    ...
```

Reply:
👍  I decided to move the GrantExchange logic into the `create_sentry_app_installation` factory